element_array_ephys/ephys_no_curation.py

```diff
@@ -1168,177 +1168,245 @@ class Waveform(dj.Part):
-            unit_templates = we_kilosort.get_all_templates()
-            unit_waveforms = np.reshape(
-                unit_templates,
-                (
-                    unit_templates.shape[1],
-                    unit_templates.shape[3],
-                    unit_templates.shape[2],
-                ),
+            unit_peak_waveform = []
+            unit_electrode_waveforms = []
+
+            for unit in units:
+                unit_peak_waveform.append(
+                    {
+                        **unit,
+                        "peak_electrode_waveform": we.get_template(
+                            unit_id=unit["unit"], mode="average", force_dense=True
+                        )[:, unit_id_to_peak_channel_indices[unit["unit"]][0]],
+                    }
+                )
+
+                unit_electrode_waveforms.extend(
+                    [
+                        {
+                            **unit,
+                            **e,
+                            "waveform_mean": mean_waveforms[
+                                unit["unit"], :, e["electrode"]
+                            ],
+                        }
+                        for e in electrode_info
+                    ]
+                )
+
+            self.insert1(key)
+            self.PeakWaveform.insert(unit_peak_waveform)
+            self.Waveform.insert(unit_electrode_waveforms)
+
+        else:
+            kilosort_dataset = kilosort.Kilosort(output_dir)
+
+            acq_software, probe_serial_number = (
+                EphysRecording * ProbeInsertion & key
+            ).fetch1("acq_software", "probe")
+
+            # -- Get channel and electrode-site mapping
+            recording_key = (EphysRecording & key).fetch1("KEY")
+            channel2electrodes = get_neuropixels_channel2electrode_map(
+                recording_key, acq_software
             )
 
-            # Approach assumes unit_waveforms was generated correctly (templates are actually the same as mean_waveforms)
-            def yield_unit_waveforms():
-                for unit_no, unit_waveform in zip(
-                    kilosort_dataset.data["cluster_ids"], unit_waveforms
-                ):
-                    unit_peak_waveform = {}
-                    unit_electrode_waveforms = []
-                    if unit_no in units:
+            # Get all units
+            units = {
+                u["unit"]: u
+                for u in (CuratedClustering.Unit & key).fetch(
+                    as_dict=True, order_by="unit"
+                )
+            }
+
+            waveforms_folder = [
+                f for f in output_dir.parent.rglob(r"*/waveforms*") if f.is_dir()
+            ]
+
+            if (output_dir / "mean_waveforms.npy").exists():
+                unit_waveforms = np.load(
+                    output_dir / "mean_waveforms.npy"
+                )  # unit x channel x sample
+
+                def yield_unit_waveforms():
```

Comment:
@ttngu207 I didn't touch this `else` block, but why is this function duplicated?